### PR TITLE
[memory][api] Divide params errors handling and memory state handling, upgrade API docs

### PR DESCRIPTION
**Pull request by NikitaZotov**
* [x] Read PR [documentation](https://github.com/ostis-ai/sc-machine/blob/main/CONTRIBUTING.md)
* [x] Update changelog
* [x] Update documentation

### CODE REVIEW

File: sc-memory/sc-memory/sc_memory.hpp [hunk 1]
```diff
@@ -87,61 +120,407 @@ class ScMemoryContext
...
 
-  /*! Change subtype of sc-element.
-   * Return true, if there are no any errors; otherwise return false.
+  /*!
+   * @brief Changes the subtype of an sc-element.
+   *
+   * This method changes the subtype of the sc-element identified by the given sc-address.
+   *
+   * @param addr The sc-address of the sc-element to modify.
+   * @param subtype The new subtype to set.
```
Comment (MksmOrlov) [inline sc-memory/sc-memory/sc_memory.hpp:311]: Why not ScType? A sc_type

**Review by MksmOrlov — CHANGES_REQUESTED** [review]
Add breaking changes

File: sc-memory/sc-core/sc-store/sc_storage.h [hunk 2]
```diff
@@ -20,196 +26,588 @@
...
+ *       the result of the operation is not explicitly returned. Use this function
+ *       when the result is not needed.
+ * @note This function is thread-safe.
  */
 sc_addr sc_storage_node_new(sc_memory_context const * ctx, sc_type type);
 
-/*! Create new sc-link
- * @return Return sc-addr of created sc-link or empty sc-addr if sc-link wasn't created
+/*!
+ * @brief Creates a new sc-node with the specified type and returns the result of the operation.
```
[the remark this answers is not in the document]
Comment (NikitaZotov) [reply sc-memory/sc-core/sc-store/sc_storage.h:134]: This is about `sc_result * result`

File: sc-memory/sc-core/sc-store/sc_storage.h [hunk 3]
```diff
@@ -20,196 +26,588 @@
...
+ * @note This function is thread-safe.
  */
 sc_addr sc_storage_link_new(sc_memory_context const * ctx, sc_type type);
 
-/*! Create new sc-arc.
- * @param type Type of new sc-arc
- * @param beg sc-addr of begin sc-element
- * @param end sc-addr of end sc-element
+/*!
+ * @brief Creates a new sc-link with the specified type and returns the result of the operation.
```
[the remark this answers is not in the document]
Comment (NikitaZotov) [reply sc-memory/sc-core/sc-store/sc_storage.h:177]: There is too

File: sc-memory/sc-memory/sc_memory.hpp (hunk 1 above)
Comment (NikitaZotov) [reply sc-memory/sc-memory/sc_memory.hpp:311]: I fix it in another PR. where I must deprecated and clarify all C++ API methods.

File: sc-memory/sc-core/sc-store/sc_storage.h [hunk 4]
```diff
@@ -20,196 +26,588 @@
...
+/*!
+ * @brief Frees the memory occupied by a sc-element and all connected elements.
+ *
+ * This function frees the memory occupied by a sc-element identified by the provided
+ * sc-addr, along with all the connected elements (input/output sc-connectors) related to it.
+ * The operation result is stored in the provided pointer to sc-result.
+ *
+ * @param ctx A pointer to the sc-memory context that manages the operation.
+ * @param addr The sc-addr of the sc-element to be freed.
+ * @param result Pointer to a variable that will store the result of the operation.
```
[the remark this answers is not in the document]
Comment (NikitaZotov) [reply sc-memory/sc-core/sc-store/sc_storage.h:99]: I don't understand that you don't see

File: sc-memory/sc-core/sc-store/sc_storage.h [hunk 5]
```diff
@@ -20,196 +26,588 @@
...
-/*! Remove sc-element from storage
- * @param addr sc-addr of element to erase
- * @return If input params are correct and element erased, then return SC_OK;
- * otherwise return SC_ERROR
+/*!
+ * @brief Frees the memory occupied by a sc-element and all connected elements.
+ *
+ * This function frees the memory occupied by a sc-element identified by the provided
+ * sc-addr, along with all the connected elements (input/output sc-connectors) related to it.
+ * The operation result is stored in the provided pointer to sc-result.
```
[the remark this answers is not in the document]
Comment (NikitaZotov) [reply sc-memory/sc-core/sc-store/sc_storage.h:95]: What is a problem?

[the remark this answers is not in the document]
Comment (MksmOrlov) [reply sc-memory/sc-core/sc-store/sc_storage.h:103]: No pointer to sc-result

**Review by MksmOrlov — CHANGES_REQUESTED** [review]
Check unresolved

File: sc-memory/sc-core/sc-store/sc_types.h [hunk 6]
```diff
@@ -183,16 +183,21 @@ enum _sc_result
...
-  SC_RESULT_ERROR_IO,              // input/output error
   SC_RESULT_ERROR_INVALID_STATE,   // invalid state of processed object
   SC_RESULT_ERROR_NOT_FOUND,       // item not found
-  SC_RESULT_NO,                    // no any result
-  SC_RESULT_UNKNOWN,               // result unknown
+  SC_RESULT_ERROR_FULL_MEMORY,
+  SC_RESULT_ERROR_FILE_MEMORY_IO,
   SC_RESULT_ERROR_ADDR_IS_NOT_VALID,
-  SC_RESULT_ERROR_ELEMENT_IS_NOT_CONNECTOR,
+  SC_RESULT_ERROR_ELEMENT_IS_NOT_NODE,
```
[the remark this answers is not in the document]
Comment (NikitaZotov) [reply sc-memory/sc-core/sc-store/sc_types.h:194]: It was added in this version and was remove in this version. So there are no need to add it into breaking changes

File: sc-memory/sc-core/sc-store/sc_storage.c [hunk 7]
```diff
@@ -810,24 +866,24 @@ sc_addr sc_storage_arc_new(sc_memory_context const * ctx, sc_type type, sc_addr
...
+  *result = SC_RESULT_OK;
   return arc_addr;
 error:
   sc_storage_free_element(arc_addr);
   sc_monitor_release_write_n(2, beg_monitor, end_monitor);
   return SC_ADDR_EMPTY;
 }
 
-sc_uint32 sc_storage_get_element_output_arcs_count(sc_memory_context const * ctx, sc_addr addr)
+sc_uint32 sc_storage_get_element_output_arcs_count(sc_memory_context const * ctx, sc_addr addr, sc_result * result)
```
[the remark this answers is not in the document]
Comment (NikitaZotov) [reply sc-memory/sc-core/sc-store/sc_storage.c:877]: It was added in this version and was remove in this version. So there are no need to add it into breaking changes

**Review by MksmOrlov — APPROVED** [review]

**Pull request merged**